ENH: Show full PEP 457 argument lists for ufuncs

eric-wieser opened the pull request:
Describes the full argument lists of `ufunc.__call__` using [PEP 457](https://www.python.org/dev/peps/pep-0457/) syntax.

Before:

    add(x1, x2, [out])
    frexp(x, [out1, out2,])

After:

    add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True, signature, extobj)
    frexp(x[, out1, out2], /[, out=(None, None)], *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])

In future, this could be hooked up to `__text_signature__` (#8734) to make `inspect.signature` work on ufuncs (right now, this requires monkey-patching `inspect`).

<s>This needs to be checked for compatibility against sphinx/numpydoc, which might choke on this signature format</s> Done and working!

eric-wieser committed:
DOC: Show full argument lists for ufuncs

eric-wieser commented:
To make inspect.signature work, we:

* Add a trivial `__text_signature__` attribute that contains what we calculate here
* Monkey-patch as `inspect._NonUserDefinedCallables += (np.ufunc,)`

mhvk commented:
I now looked in a bit more detail, and this looks all good. It might make sense to have it in 1.13 still, but perhaps @charris should decide on that.

charris commented:
@mhvk If you think it is good, put it in. The 1.13.0 milestone is for things I think need to be in 1.13, and some of those I'll probably push off again to 1.14.

eric-wieser commented:
Note that if we do put this in, the arguments still wont be described on the individual ufunc pages - so somehow, we should probably link back to the ufunc docs

mhvk commented:
@eric-wieser - you mean that these changes are visible in, say, ipython, but won't be seen by sphinx?

eric-wieser commented:
I mean that the top of the sphinx page will list all the arguments, but nowhere on the page will a description of the arguments be given.

Having said that, I haven't actually tested that sphinx supports this syntax...

eric-wieser commented:
> Having said that, I haven't actually tested that sphinx supports this syntax...

And havehad difficulty in the past on windows getting numpy to build docs from the source, rather than the installed version

mhvk commented:
Ah, yes, so the general description has to be adapted too. In that case, my suggestion would be to merge this with 1.14 as a milestone, and raise a new issue to also adjust the overall documentation.

eric-wieser commented:
> In that case, my suggestion would be to merge this with 1.14 as a milestone

Huh? Do you mean wait until 1.13 is branched, and then merge? Or set the milestone to 1.14 yet merge it into 1.13?

eric-wieser committed:
DOC: Refer to main ufunc docs for kwargs

eric-wieser committed:
DOC: Remove explanation of exception due to out

This was covered by very few ufuncs, and is probably better to leave out.

eric-wieser commented:
> Ah, yes, so the general description has to be adapted too. 

Done. All this needs is someone with a working doc build setup to check that the pages for `np.add` and `np.frexp` look right under sphinx.

eric-wieser commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -10,13 +10,29 @@
+        Alternate array object(s) in which to put the result; if provided, it
+        must have a same shape that the inputs broadcast to.
+    where : array_like, optional
+        Values of True indicate to calculate the ufunc at that position, values
```
As a bonus, this now means that `where` is documented everywhere.

We could add the others here, but the explanations are more complicated

eric-wieser commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -10,13 +10,29 @@
+""").strip()
+
 def add_newdoc(place, name, doc):
+    doc = textwrap.dedent(doc).strip()
```
dedenting here makes substituting `$PARAMS` do the right thing

mhvk commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -10,13 +10,29 @@
+    where : array_like, optional
+        Values of True indicate to calculate the ufunc at that position, values
+        of False indicate to leave the value in the output alone.
+
```
To make it clearer that there are other arguments, I'd put this as part of the parameter list:
```
**kwargs
    For other keyword-only ...
```

eric-wieser replied:
Done

mhvk commented:
For reasons slightly beyond me, I have similar problems with the sphinx build using the system installation of numpy despite being in a virtual environment in which I have installed the PR. So, I cannot immediate check that things work there.

Inside ipython, things look good, except I think one might as well adjust the class docstring as well...

eric-wieser committed:
DOC: Simplify the output of help(ufunc)

The distinction between unary and binary is not too helpful, and ignores the fact that trinary ufuncs can exist.

Also add a link to the ufunc docs

eric-wieser commented:
Ok, this can be tested with `runtests.py --shell`, and then building the docs from there.

The signature works just fine, but the link doesn't work:

![image](https://cloud.githubusercontent.com/assets/425260/25670517/d6ae7c8e-3024-11e7-935d-5cc219779015.png)

![image](https://cloud.githubusercontent.com/assets/425260/25670541/ec561f42-3024-11e7-800a-d681fea3d561.png)

pv commented:
I think numpydoc doesn't like the link in the parameter list, as there 
was no allowance in the spec for stuff after the parameter list.
.
The extra explanation can probably be put in a Notes section or so.

eric-wieser commented:
@pv: I'm running into #9046 here, which means those screenshots do not match my source

eric-wieser commented:
Ok, after doing a truly clean build, this now works:

![image](https://cloud.githubusercontent.com/assets/425260/25671436/f5972062-3027-11e7-9905-0a1877daf923.png)

eric-wieser commented:
Arguments for putting this in 1.13: We're now allowing people to intercept all the ufunc arguments, so it'd be handy to give people a reminder of what they are, without having to resort to the online docs

mhvk commented:
I think you ended up making the full change, right? If so, I think 1.13 is indeed appropriate.

mhvk commented on numpy/add_newdocs.py:
```diff
@@ -5421,40 +5421,19 @@ def luf(lamdaexpr, *args, **kwargs):
 
-    To see the documentation for a specific ufunc, use np.info().  For
-    example, np.info(np.sin).  Because ufuncs are written in C
+    To see the documentation for a specific ufunc, use `np.info`.  For
```
I think this needs to be double ticks to, or does it link correctly? (I ask since in astropy this would require `~numpy.info`; if the link does work, maybe pre-pend :func:)

eric-wieser replied:
Forgot to check this. I think you're probably right there.

eric-wieser replied:
This page never gets generated anyway, so this is kinda irrelevant

eric-wieser replied:
Switched to `info` for simplicity, in case this does get linked in future

mhvk commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -10,13 +10,30 @@
+
 def add_newdoc(place, name, doc):
+    doc = textwrap.dedent(doc).strip()
+    doc = doc.replace('$PARAMS', _params_text)
```
Bit odd to have this shell-like contruct. Why not just call it `{PARAMS}` in the strings, and here do
```
doc = doc.format(PARAMS=_params_text)
```

eric-wieser replied:
Latex in the docstrings was my concern here

mhvk replied:
OK, that seems fair enough. Just leave as is, then.

mhvk commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -3067,6 +3083,7 @@ def add_newdoc(place, name, doc):
     ----------
     x1 : array_like
```
Might as well change this to `x`

eric-wieser replied:
Yep, since right now it doesn't actually match the signature

eric-wieser replied:
Done

mhvk reviewed (commented):
Had another look and found a few small things to change, but overall, this looks great!

mhvk commented:
I think after #9063 makes most sense, but I'm changing the milestone so this won't be forgotten!

eric-wieser committed:
DOC: Correct parameter names for spacing and invert

eric-wieser commented:
@mhvk: Rebased to include the modifications to the `divmod` docstring

charris commented on numpy/add_newdocs.py:
```diff
@@ -5463,18 +5442,22 @@ def luf(lamdaexpr, *args, **kwargs):
-    out : array_like
-        An array to store the output. Must be the same shape as the
-        output would have.
+    *x : array_like
```
Each of the  `*` need escaping like so `\\*` for sphinx.  If you make it a raw string you can use a single `\'.

eric-wieser replied:
This file never actually reaches sphinx anyway, but I was under the impression that this worked correctly due to the transformation that the `Parameters` section undergoes

charris replied:
I've had to fix other uses before as the docs didn't render correctly.

eric-wieser replied:
Well, this doesn't even get rendered, so no danger of it getting rendered incorrectly!

Since the only place this is ever seen is in the terminal, I'd argue it's better without backslashes.

I'd also argue that to keep the terminal tidy, the fix should go in the numpy sphinx extension, not here

charris replied:
I agree that having numpydoc escape the `*`'s when generating the sphinx input would be nice, but it doesn't at this time. If this never showed up in the documentation, that would not be a problem, but I assume you would like it to eventually show up.

eric-wieser replied:
> but I assume you would like it to eventually show up.

Not really - AFAICT, the sole purposed of this page is to redirect the user to `np.info` or the ufunc docs. if `ufunc.__doc__` is made visible in the web docs, then IMO, it should be the contents of `ufuncs.rst`.

This docstring aside - I was pretty sure that `**kwargs` was parsed correctly in the other docstring modified by this changeset - does the issue only exist for single asterisks?

charris commented on numpy/add_newdocs.py:
```diff
@@ -5463,18 +5442,22 @@ def luf(lamdaexpr, *args, **kwargs):
+    where : array_like, optional
+        Values of True indicate to calculate the ufunc at that position, values
+        of False indicate to leave the value in the output alone.
+    **kwargs :
```
Don't know if this needs something after the `:` or not, but might be safer to put a placeholder there.

eric-wieser replied:
That `:` should probably be removed, but again - no HTML docs are generated from this page

mhvk replied:
Irrelevant given that there are documents generated, but yes, with the : removed it works fine.

eric-wieser replied:
Fixed, anyway

charris commented:
> This docstring aside - I was pretty sure that **kwargs was parsed correctly in the other docstring modified by this changeset

Not sure. The `*` gets translated as emphasis IIRC. Grepping the source, the two versions I see are `\\*\\*kwargs` or just plain `kwargs` without the asterics. The latter seems like a viable workaround as long as it comes with an `optional arguments` explanation or some such. The same variations hold for `args`. I suspect a fix for numpydocs would not be too difficult if one finds where the parameters are parsed out.

eric-wieser commented:
Ok, see the screenshot in the above comment - the `**` is definitely handled correctly.

mhvk commented:
`**kwargs` works fine -- see also http://docs.astropy.org/en/latest/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord (astropy uses the numpy setup with some additions)

eric-wieser commented:
@mhvk: Do you have an example of *args working too, to save me the trouble of a local build?

eric-wieser commented:
@charris: tested with unescaped `*` in the arglist, and it works just fine (sphinx 1.5)

charris commented:
Looks like the escape problem has been fixed, good to know. Now we can clean up the other cases ;)

charris commented:
Thanks Eric.
